### ISSUE TITLE
Learn the inverse of 2D function f : 2D -> 2D

### ISSUE THREAD

**Issue by ucacaxm** [Microsoft/CNTK#717]
Hi,

I would like to train a network for inverting a function with 2 inputs and 2 ouputs.
I started with a simple case of my problem: f(a,b) -> (x,y)
x = cos(a) +cos(a+b))
y = sin(a) +sin(a+b))

a, b are between [-PI, + PI]  and the ouput between [-2, +2]

I would like to learn the inverse of that function. f-1(x,y) -> (a,b)

When I do it in 1D the error prediction is 0 (which means it learns).
but I do not manage to get better result than 'error prediction' of 0.5 ...
I tried to designed 2 network and concatenate results at the end with Splice.

Any idea ? I attach the BS script at the end of the post.
Maybe my problem is around the normalization ?
I did that 
        fmean = Mean(features);
        finvstd=InvStdDev(features)
        finput=PerDimMeanVarNormalization(features, fmean, finvstd)
but do I need to do something at the end ?

Thank you
Alex
###### 
# TRAINING CONFIG
###### 

IK_TrainXY = [
    action = "train"

```
# network description
BrainScriptNetworkBuilder=[

    include "ik.bs"

    # sample and label dimensions
    featuresDim = 2
    labelDim = 2        
    hiddenDim = 200

    features = Input (featuresDim)
    labels   = Input (labelDim)

    fmean = Mean(features);
    finvstd=InvStdDev(features)
    finput=PerDimMeanVarNormalization(features, fmean, finvstd)

    x1 = DNNLayer ( featuresDim,   hiddenDim, finput)
    x  = DNNLayer     ( hiddenDim, 1,  x1)

    y1 = DNNLayer ( featuresDim,   hiddenDim, finput)
    y  = DNNLayer     ( hiddenDim, 1,  y1)

    xy = Splice ((x:y), axis=1)

    # Criterion
    criterion = SquareError(labels, xy)
    # Eval
    errorPred = ErrorPrediction(labels, xy)


    # root nodes
    featureNodes    = (features)
    labelNodes      = (labels)
    criterionNodes  = (criterion)
    evaluationNodes = (errorPred)
    outputNodes     = (xy)
]


SGD = [
    # epochSize = 0 means epochSize is the size of the training set         
    epochSize = 0                   # =0 means size of the training set
    minibatchSize = 25
    learningRatesPerSample = 0.004   # gradient contribution from each sample
    momentumAsTimeConstant = 0
    maxEpochs = 50  
]




# Parameter values for the reader       
reader = [
    readerType = "CNTKTextFormatReader"
    file = "$DataDir$/ikDataTrainXYinv.txt"
    input = [
        features = [
            dim = 2
            format = "dense"
        ]
        labels = [
            dim = 2
            labelType = "Regression"
            format = "dense"
        ]
    ]
]
```

]

**Comment by frankseide** [comment]
Doing it jointly should always work better, unless the two dimensions are effectively independent (x=f(a), y = g(b)). With only one hidden layer, the difference might be small, but I would still recommend to do that.

The reason you don't see a proper error rate I think is that `ErrorPrediction()` is not the correct criterion. It expects `labels` to be one-hot, which is not the case in your application. In your application, I would use the same criterion for training and evaluation.

The correct error metric depends on your use case. `SquareError()` may be the right one, unless some value ranges are more critical than others (but you will get some of that from the distribution of your training data).

The normalization is correct. This normalization only affects convergence, but assume you find the global optimum with and without normalization, the resulting networks should give identical outputs.

**Comment by benfre** [comment]
From Wolfram Alpha, except few special cases, there are two general results:

http://www.wolframalpha.com/input/?i=Solve%5Bx+%3D%3D+Cos%5Ba%5D+%2B+Cos%5Ba+%2B+b%5D+%26%26+y+%3D%3D+Sin%5Ba%5D+%2B+Sin%5Ba+%2B+b%5D,+%7Ba,+b%7D%5D

**Comment by ucacaxm** [comment]
Thank for your reply.

frankseide, you advise me to do something like that

		h1 = DNNLayer ( featuresDim,   hiddenDim, finput)
		xy = DNNLayer ( hiddenDim, labelDim, h1)
		
		
		# Criterion
		criterion = SquareError(labels, xy)
		# Eval
		errorPred = SquareError(labels, xy)
		
		
        # root nodes
        featureNodes    = (features)
        labelNodes      = (labels)
        criterionNodes  = (criterion)
        evaluationNodes = (errorPred)
        outputNodes     = (xy)

The results are these:
Starting Epoch 50: learning rate per sample = 0.004000  effective momentum = 0.000000  momentum as time constant = 0.0 samples
BlockRandomizer::StartEpoch: epoch 49: frames [490000..500000] (first sequence at sample 490000), data subset 0 of 1

Starting minibatch loop.
 Epoch[50 of 50]-Minibatch[   1-  10, 2.50%]: criterion = 0.54975336 * 250; errorPred = 0.54975336 * 250; time = 0.0941s; samplesPerSecond = 2655.4
...
 Epoch[50 of 50]-Minibatch[ 391- 400, 100.00%]: criterion = 0.50580859 * 250; errorPred = 0.50580859 * 250; time = 0.1039s; samplesPerSecond = 2406.2
Finished Epoch[50 of 50]: [Training] criterion = 0.56601216 * 10000; errorPred = 0.56601216 * 10000; totalSamplesSeen = 500000; learningRatePerSample = 0.0040000002; epochTime=3.84602s


Maybe it is something with in the SGD parameters ?
Or maybe I misunderstand the results:
errorPred = 0.50580859 * 250
For a function like that, the errorPred should be 0,  ?


benfre, I know that an analytic solution might exist. It is mean that a DNN should learn that easily.
For my application later I will do something similar with a dimension of 50 or 100 as input.

**Comment by benfre** [comment]
I mean the inverse mapping is not one to one. the training input will have same features but different label, and the test will have the same problem. That is why the error rate may never go down.
